### PR TITLE
feat: detach with machine storage

### PR DESCRIPTION
**Pull request by SimonRichardson**
When reprovisioning a machine whose backing instance has been lost, any machine-scoped storage attached to that instance must also be treated as lost. Reusing its provider IDs, attachment plans, or block-device records could make the replacement machine appear to have storage that no longer exists.

Previously, the detach operation cleared the old machine and network provider data but retained storage-related state. It did not classify attached storage or reset its physical provider state. This meant machine-scoped storage could retain information belonging to the lost instance, while model-scoped storage did not prevent reprovisioning.

Reprovisioning now classifies attached volumes and filesystems before proceeding. Model-scoped storage is rejected because safely reattaching storage that may outlive the machine requires a separate recovery flow. Ambiguous, inconsistent, or non-alive storage also fails closed.

For machine-scoped storage, Juju preserves its identity and logical attachment intent while clearing provider-specific state, attachment plans, and old block-device evidence. The storage is returned to pending, allowing normal provisioning to create empty replacement storage for the new instance.

Storage is checked again within the detach transaction so that changes made after the initial eligibility check cannot bypass these safeguards.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model foo
```

Test machine-scoped storage:

```sh
$ juju deploy juju-qa-dummy-storage machine-storage \
  --storage single-fs=rootfs
$ juju storage
```

Stop the Juju services on the machine and delete its backing LXD instance.

```sh
# stop the instance manually
$ juju add-ssh-key <your ssh key>
$ juju ssh 0
$ sudo systemctl stop juju*.service
```

```sh
$ lxc delete <instance id> -f
```

Now retry provisioning:

> [!NOTE]
> This won't create a new instance, as we've not wired up the provisioner task to spot this. So for now you should see that the `instance_id` in the `juju_db_repl` for the `machine_cloud_instance` is `<nil>`.

```sh
$ juju reprovision-machine 0
```

### Model scoped storage

```
$ juju add-model model-storage
$ juju deploy juju-qa-dummy-storage \
  --storage single-fs=lxd
```

Stop the Juju services and delete the backing LXD instance, then run:

```
$ juju reprovision-machine 0
ERROR reprovisioning machine "0": model-scoped storage attached to machine
```

## Links

**Jira card:** [JUJU-10181](https://warthogs.atlassian.net/browse/JUJU-10181)

[JUJU-10181]: https://warthogs.atlassian.net/browse/JUJU-10181?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/machine/state/reprovision.go [hunk 1]
```diff
@@ -346,3 +379,407 @@ func runReprovisionStatements(
...
+	volumes, filesystems                map[string]struct{}
+	plans, volumeAttachments            map[string]struct{}
+	filesystemAttachments, blockDevices map[string]struct{}
+}
+
+func (st *State) prepareReprovisionStorageTargetStatements() (
+	*sqlair.Statement, *sqlair.Statement, error,
+) {
+	volumeStmt, err := st.Prepare(`
+WITH volume_targets AS (
```
Comment (nvinuesa) [inline domain/machine/state/reprovision.go]: Aren't we missing `machine_volume`? Is this intended?

Comment (SimonRichardson) [reply domain/machine/state/reprovision.go]: Yeah, that was a mistake. Storage is a complex beast.

File: domain/machine/state/reprovision.go [hunk 2]
```diff
@@ -346,3 +379,407 @@ func runReprovisionStatements(
...
+        ON sv.uuid = siv.storage_volume_uuid
+    LEFT JOIN storage_instance AS si
+        ON siv.storage_instance_uuid = si.uuid
+    LEFT JOIN storage_volume_attachment AS sva
+        ON sv.uuid = sva.storage_volume_uuid
+        AND sva.net_node_uuid = $reprovisionStorageTargetParams.net_node_uuid
+    LEFT JOIN storage_volume_attachment_plan AS svap
+        ON sv.uuid = svap.storage_volume_uuid
+        AND svap.net_node_uuid = $reprovisionStorageTargetParams.net_node_uuid
+    WHERE sva.uuid IS NOT NULL OR svap.uuid IS NOT NULL
```
Comment (jack-w-shaw) [inline domain/machine/state/reprovision.go]: P1: This seed condition misses a volume that is still owned by this machine through `machine_volume` but currently has neither an attachment nor an attachment plan. In that state the detach proceeds without classifying or resetting the volume, so its old `provider_id`, physical attributes, and status can survive the loss of the backing instance. The filesystem query handles the equivalent case by seeding from `machine_filesystem` and then classifying a missing attachment as ambiguous. Please seed volume targets from `machine_volume` as well (alongside attachment/plan targets), and add the corresponding missing-attachment regression test.

**Review by jack-w-shaw — APPROVED** [review]
Approving overall, with one important storage-classification edge case noted inline.

**Comment by SimonRichardson** [comment]
@nvinuesa @jack-w-shaw I've reworked the statements now, it should be a lot simpler. Now we do the life check up front, which makes the rest of the queries simpler.

**Review by nvinuesa — APPROVED** [review]
Queries look way better now, thanks for the updates.
